### PR TITLE
tracing: proper child span support in gRPC client

### PR DESCRIPTION
**Pull request by mattklein123**
- Remove finalizer concept as it isn't needed and requires extra
  allocations.
- Plumb tracing through into gRPC client and wire up in ratelimit
  client.
- Do not require null allocation for span in connection manager
  if tracing is not configured.
- Remove propagating x-request-id for ratelimit requests as we
  now have proper tracing support. If this is requested in the
  future we should make request ID a part of Tracing::Span
  itself.

Signed-off-by: Matt Klein <mklein@lyft.com>

### CODE REVIEW

**Comment by mattklein123** [comment]
@objectiser @goaway @rshriram 

@objectiser this will conflict with your PR I think. We can merge yours first.

File: source/common/grpc/async_client_impl.h [hunk 1]
```diff
@@ -257,17 +259,47 @@ class AsyncStreamImpl : public AsyncStream<RequestType>,
...
+public:
+  // Tracing::Config
+  Tracing::OperationName operationName() const override { return Tracing::OperationName::Egress; }
+  const std::vector<Http::LowerCaseString>& requestHeadersForTags() const override {
+    return request_headers_for_tags_;
+  }
+
+  struct {
+    const std::string STATUS = "status";
+    const std::string GRPC_STATUS = "grpc-status";
```
[the remark this answers is not in the document]
Comment (mattklein123) [reply source/common/grpc/async_client_impl.h]: sure will do

File: source/common/grpc/async_client_impl.h [hunk 2]
```diff
@@ -278,11 +310,17 @@ class AsyncRequestImpl : public AsyncRequest,
   }
 
   // Grpc::AsyncRequest
-  void cancel() override { this->resetStream(); }
+  void cancel() override {
+    current_span_->setTag(TracingConfig::get().TagStrings.STATUS,
+                          TracingConfig::get().TagStrings.CANCELED);
```
[the remark this answers is not in the document]
Comment (mattklein123) [reply source/common/grpc/async_client_impl.h]: I thought about this, but my thinking is that this is a local cancel, not a remote cancel, so we shouldn't confuse it with gRPC remote cancellation?

File: source/common/grpc/async_client_impl.h [hunk 3]
```diff
@@ -295,20 +333,26 @@ class AsyncRequestImpl : public AsyncRequest,
...
+
     if (status != Grpc::Status::GrpcStatus::Ok) {
-      callbacks_.onFailure(status, message);
-      return;
-    }
-    if (response_ == nullptr) {
-      callbacks_.onFailure(Status::Internal, EMPTY_STRING);
-      return;
+      current_span_->setTag(TracingConfig::get().TagStrings.ERROR,
+                            TracingConfig::get().TagStrings.TRUE);
```
[the remark this answers is not in the document]
Comment (mattklein123) [reply source/common/grpc/async_client_impl.h]: We do store the numeric status above. Are you suggesting a new tag name with spelled out status?

File: source/common/grpc/async_client_impl.h (hunk 2 above)
[the remark this answers is not in the document]
Comment (objectiser) [reply source/common/grpc/async_client_impl.h]: Good point - best to keep separate.

File: source/common/grpc/async_client_impl.h (hunk 3 above)
[the remark this answers is not in the document]
Comment (objectiser) [reply source/common/grpc/async_client_impl.h]: No, ignore - storing the integer is fine.

**Comment by mattklein123** [comment]
@objectiser updated

**Comment by goaway** [comment]
LGTM. Basically what I was going for sans the finalizer overhead, which I agree ends up being an improvement.

File: source/common/grpc/async_client_impl.h [hunk 4]
```diff
@@ -257,17 +259,47 @@ class AsyncStreamImpl : public AsyncStream<RequestType>,
...
-                   AsyncRequestCallbacks<ResponseType>& callbacks,
+                   AsyncRequestCallbacks<ResponseType>& callbacks, Tracing::Span& parent_span,
                    const Optional<std::chrono::milliseconds>& timeout)
       : AsyncStreamImpl<RequestType, ResponseType>(parent, service_method, *this, timeout),
-        request_(request), callbacks_(callbacks) {}
+        request_(request), callbacks_(callbacks) {
+
+    current_span_ = parent_span.spawnChild(TracingConfig::get(),
+                                           "async " + parent.remote_cluster_name_ + " egress",
+                                           ProdSystemTimeSource::instance_.currentTime());
```
Comment (danielhochman) [inline source/common/grpc/async_client_impl.h]: reminds me that we need to upgrade lightstep to add monotonic time to tracing

File: source/common/grpc/async_client_impl.h [hunk 5]
```diff
@@ -257,17 +259,47 @@ class AsyncStreamImpl : public AsyncStream<RequestType>,
   friend class AsyncClientImpl<RequestType, ResponseType>;
 };
 
+class AsyncClientTracingConfig : public Tracing::Config {
+public:
+  // Tracing::Config
+  Tracing::OperationName operationName() const override { return Tracing::OperationName::Egress; }
```
Comment (danielhochman) [inline source/common/grpc/async_client_impl.h]: missing cov for :265-:268

File: source/common/grpc/async_client_impl.h (hunk 4 above)
Comment (mattklein123) [reply source/common/grpc/async_client_impl.h]: Yeah we do need to do this. I think that entire driving is being worked on so hopefully we can get it swapped out with the OT driver update.

**Comment by mattklein123** [comment]
@danielhochman @goaway updated

**Pull request merged**